Replace queue-based interrupts with NVIC regs

alevy opened the pull request:
### Pull Request Overview

Instead of using a global queue to notify the event loop about pending
interrups, this commit uses the built in NVIC pending bits themselves.
This has a few benefits:

1. It removes controller specific ISRs that do nothing but enqueue an interrupt event.
2. It makes this code generic for all Cortex-M4 chips (and should be ported to Cortex-M0 et
al). As a result of both of these, it removes about 10kB of code on hail
and imix.
3. it removes the need for controller drivers to handle
enabling/disabling nvic lines themselves.
4. it's much faster since enqueing and dequeuing in the ring buffer is actually pretty expensive.

Downsides?

* The nvic indexes can't be an enum anymore since those "names" are chip specific. Oh well...

### Testing Strategy

~~In principle, this should be a pretty transparent change except some drivers relied on explicit nvic handling to disable interrupts. I believe I've fixed that in all cases, and have tested a number of the example apps on hail and imix. However, more comprehensive testing would be useful.~~

Tested fairly extensively on Hail, NRF51-DK and NRF52-DK. On Hail, we've run a bunch of apps, including `tests/hail`, `rot13`, `test/nonvolatile_storage`, `blink` and on the NRFs we ran `ble_advertising`, `blink`, `tests/aes`.

### TODO or Help Wanted

~~The flash driver has a particular place where it actually called the ISR explicitly from the code. That ISR no longer exists (and either way would be useless) and I don't understand what role exactly that was playing, or how to reproduce that functionality (perhaps by just calling `handle_interrupt`, but why did the original code not just do that?).~~

~~I left a TODO where that code used to be. Does anyone know how to trigger the issue that led to that hack?~~

### Documentation Updated

- [x] ~~Kernel: The relevant files in `/docs` have been updated or no updates are required.~~
- [x] ~~Userland: The application README has been added, updated, or no updates are required.~~

### Formatting

- [x] `make formatall` has been run.

bradjc commented:
Oh I also tested with the blink, hail, and rot13 apps all running concurrently and they seemed to work.

alevy commented:
Is this intentionally dated for tomorrow instead of today?

ppannuto commented on arch/cortex-m4/src/lib.rs:
```diff
@@ -66,30 +66,37 @@ pub unsafe extern "C" fn generic_isr() {
+
+    /* Set thread mode to privileged */
+    mov r0, #0
+    msr CONTROL, r0
```
`generic_isr` is only called to service interrupts, correct? My reading of B1.3.1 of the ARM v7m ARM is that the processor should be in Handler mode, which is by definition already privileged. I don't think this is needed.

alevy replied:
This is needed to make sure we're in privileged mode when we return the kernel, not for this handler's code.

ppannuto commented on arch/cortex-m4/src/lib.rs:
```diff
@@ -66,30 +66,37 @@ pub unsafe extern "C" fn generic_isr() {
 
-    ldr r1, =INTERRUPT_TABLE
-    ldr r0, [r1, r0]
+    /* r1 = NVIC.ICER[r0 / 32] */
```
From this comment, I expect one of the instructions below to be a memory read?

alevy replied:
OK, it's taking an address: `r1 = &NVIC.ICER[r0 / 32]`

ppannuto commented on arch/cortex-m4/src/lib.rs:
```diff
@@ -66,30 +66,37 @@ pub unsafe extern "C" fn generic_isr() {
+    movt r1, #0xe000
+    lsr r2, r0, #5
+    mov r3, #4
+    mul r2, r3
```
Why load 4 into a register and multiply instead of just shifting again here?

alevy replied:
cause like... I don't know how to do that... would it just be `lsr r2, r0, #7`?

ppannuto commented on arch/cortex-m4/src/lib.rs:
```diff
@@ -66,30 +66,37 @@ pub unsafe extern "C" fn generic_isr() {
+    /* r2 = 1 << (r0 & 31) */
+    mov r2, #1
+    mov r3, #31
+    and r3, r0
```
You can consolidate these two instructions to just
`and r3, r0, #31`

ppannuto commented on arch/cortex-m4/src/nvic.rs:
```diff
@@ -0,0 +1,110 @@
+use kernel::common::volatile_cell::VolatileCell;
+
+#[repr(C, packed)]
+// Registers for the NVIC. Each
```
`Each` what? Don't leave me hanging! :D

immesys replied:
You get the answer when you buy the DLC

alevy commented on arch/cortex-m4/src/nvic.rs:
```diff
@@ -0,0 +1,110 @@
+//! Cortex-M3 NVIC
```
That's just... like... your opinion man

alevy commented:
Rebased, fixedup, ported to NRF52, cortex-m0/nrf51, but need to test after the rebase and some cleanup.

Last I tested thoroughly before the rebase, sam4l was working as was nrf52. More tomorrow

niklasad1 commented:
Looks good, have you tested it on both nRF51-DK and nRF52-DK?

alevy commented:
@niklasad1 literally just tested right now, and after some fixes, it's all working.

The one thing i noticed is that `ble_passive_scan` seems to block after a few scans, but that is happening on master as well.

alevy commented:
I think this is ready, maybe it could use some more assembly auditing from the venerable @ppannuto. I trust the assembly insofar as I've tested it, but @ppanuto's caught obvious things in the past that I've missed, and is also good at asking for the right documentation for assembly.

bradjc commented on arch/cortex-m4/src/nvic.rs:
```diff
@@ -0,0 +1,110 @@
+//! Cortex-M0 NVIC
```
M0 now?

alevy replied:
Gah! Too many files!!!

bradjc commented on boards/hail/src/main.rs:
```diff
@@ -69,6 +69,7 @@ struct Hail {
     crc: &'static capsules::crc::Crc<'static, sam4l::crccu::Crccu<'static>>,
     dac: &'static capsules::dac::Dac<'static>,
     aes: &'static capsules::symmetric_encryption::Crypto<'static, sam4l::aes::Aes>,
+    stfu_holding: &'static capsules::nonvolatile_storage_driver::NonvolatileStorage<'static>,
```
We don't want this in Hail proper.

bradjc commented on chips/sam4l/src/flashcalw.rs:
```diff
@@ -317,6 +317,7 @@ impl FLASHCALW {
             FlashState::WriteUnlocking { page } => {
                 self.current_state.set(FlashState::WriteErasing { page: page });
                 self.flashcalw_erase_page(page);
+                DEFERED_CALL.set();
```
Don't need this

alevy replied:
OK, removed it and I trust you, but I needed it before your changes on master, so should test this just in case before merging (which I can do first thing tomorrow)

alevy replied:
Tested. Works.

ppannuto commented on arch/cortex-m4/src/lib.rs:
```diff
@@ -66,38 +64,37 @@ pub unsafe extern "C" fn generic_isr() {
+    /* NVIC.ICER[r0 / 32] = 1 << (r0 & 31) */
+	movs r2, #1
+	movs r3, #32
+	sdiv r3, r0, r3
```
Replace these two with `asr r3, r0, #5`? Shorter and matches m0 implementation

ppannuto commented on arch/cortex-m0/src/lib.rs:
```diff
@@ -3,7 +3,75 @@
+    /* NVIC.ICER[r0 / 32] = 1 << (r0 & 31) */
+	movs	r2, #31
+	asrs	r3, r0, #31
+	ands	r3, r2
```
Now I'm a bit more confused, as it looks like this is and'ing with the shifted r0? I think commenting each step will alleviate confusion here.

alevy commented:
@ppannuto I cleaned up and commented the assembly for disabling nvics in the ISR.

I think the problem was that I had compiled the assembly from C code that used signed (rather than unsigned) integers so GCC was be overly clever. I re-did it with unsigned ints and the result was good enough for me to understand and comment.

ppannuto commented on chips/sam4l/src/flashcalw.rs:
```diff
@@ -67,6 +67,8 @@ enum RegKey {
     GPFRLO,
 }
 
+static DEFERED_CALL: DeferedCall = unsafe { DeferedCall::new(Task::Flashcalw) };
```
This is pervasive -- and don't hate me -- s/defered/deferred

ppannuto commented on chips/sam4l/src/dma.rs:
```diff
@@ -200,14 +194,12 @@ impl DMAChannel {
 
     pub fn handle_interrupt(&mut self) {
         let registers: &mut DMARegisters = unsafe { mem::transmute(self.registers) };
+        registers.interrupt_disable.set(0xffffffff);
```
nit: you use `!0` pretty much everywhere else

alevy commented:
@ppannuto addressed your comments. I think also all of @bradjc's comments are addressed.

(Fine if approving/merging this waits until after Thanksgiving break)

niklasad1 commented:
@alevy yeah, I have noticed  that the ```ble_passive_scan``` app hangs too.

But I haven't understood how to debug user-space applications so I have `ignored` it but perhaps best to create an issue! I thought it run out of memory but I don't know!
